src/clp_ffi_py/ir/native/PyDeserializer.cpp

```diff
@@ -0,0 +1,286 @@
+    m_deserializer_buffer_reader = DeserializerBufferReader::create(input_stream, buffer_capacity);
+    if (nullptr == m_deserializer_buffer_reader) {
+        return false;
+    }
+
+    try {
+        PyDeserializer::IrUnitHandler::LogEventHandle log_event_handle
+                = [this](clp::ffi::KeyValuePairLogEvent&& kv_log_event) -> IRErrorCode {
+            return this->handle_log_event(std::move(kv_log_event));
+        };
+
+        PyDeserializer::IrUnitHandler::UtcOffsetChangeHandle trivial_utc_offset_handle
+                = []([[maybe_unused]] clp::UtcOffset, [[maybe_unused]] clp::UtcOffset
+                  ) -> IRErrorCode { return IRErrorCode::IRErrorCode_Success; };
+
+        PyDeserializer::IrUnitHandler::SchemaTreeNodeInsertionHandle
+                trivial_schema_tree_node_insertion_handle
+                = []([[maybe_unused]] clp::ffi::SchemaTree::NodeLocator) -> IRErrorCode {
+            return IRErrorCode::IRErrorCode_Success;
+        };
+
+        PyDeserializer::IrUnitHandler::EndOfStreamHandle end_of_stream_handle
+                = [this]() -> IRErrorCode { return this->handle_end_of_stream(); };
+
+        auto deserializer_result{Deserializer::create(
+                *m_deserializer_buffer_reader,
+                {std::move(log_event_handle),
+                 std::move(trivial_utc_offset_handle),
+                 std::move(trivial_schema_tree_node_insertion_handle),
+                 std::move(end_of_stream_handle)}
+        )};
+        if (deserializer_result.has_error()) {
+            PyErr_Format(
+                    PyExc_RuntimeError,
+                    cDeserializerCreateErrorFormatStr.data(),
+                    deserializer_result.error().message().c_str()
+            );
+            return false;
+        }
+        m_deserializer = new clp::ffi::ir_stream::Deserializer<PyDeserializer::IrUnitHandler>{
+                std::move(deserializer_result.value())
+        };
+    } catch (clp::TraceableException& exception) {
+        handle_traceable_exception(exception);
+        return false;
+    }
+
+    return true;
+}
+
+auto PyDeserializer::deserialize_log_event() -> PyObject* {
+    try {
+        while (false == is_stream_completed()) {
+            auto const ir_unit_type_result{
+                    m_deserializer->deserialize_next_ir_unit(*m_deserializer_buffer_reader)
+            };
+            if (ir_unit_type_result.has_error()) {
+                if (false == handle_incomplete_ir_error(ir_unit_type_result.error())) {
+                    return nullptr;
+                }
+                break;
+            }
+            if (IrUnitType::LogEvent != ir_unit_type_result.value()
+                || false == has_unreleased_deserialized_log_event())
+            {
+                continue;
+            }
+            return py_reinterpret_cast<PyObject>(
+                    PyKeyValuePairLogEvent::create(release_deserialized_log_event())
+            );
```

Comment:
Actually it might be better if the return IR unit type is checked. Re-write it this way, do u think it's better? @haiqi96